File: apps/admin/src/services/phase2Dashboard.js
const RISK_LABELS = {
  Low: '低风险',
  Medium: '中风险',
  High: '高风险',
  Forbidden: '禁用',
};

const STATUS_LABELS = {
  queued: '排队中',
  running: '运行中',
  succeeded: '成功',
  failed: '失败',
  retry_pending: '待重试',
  manual_review: '人工复核',
  blocked: '已阻断',
  open: '待处理',
};

function toNumber(value) {
  const number = Number(value);
  return Number.isFinite(number) ? number : 0;
}

function moneyText(value, currency = 'CNY') {
  const amount = toNumber(value).toFixed(2);
  if (currency === 'CNY' || currency === 'RMB') {
    return `¥${amount}`;
  }
  return `${currency} ${amount}`;
}

function percentOf(current, limit) {
  const currentNumber = toNumber(current);
  const limitNumber = toNumber(limit);
  if (limitNumber <= 0) {
    return 0;
  }
  return Math.min(100, Math.round((currentNumber / limitNumber) * 100));
}

function riskClass(riskLevel) {
  if (riskLevel === 'Low') return 'green';
  if (riskLevel === 'Medium') return 'amber';
  return 'red';
}

function normalizeTaskRun(item = {}, currency = 'CNY') {
  const taskType = item.task_type || item.taskType || 'UNKNOWN';
  const status = item.status || 'unknown';
  const costCurrency = item.cost_currency || item.costCurrency || currency;
  return {
    runId: item.agent_task_run_id || item.agentTaskRunId || '',
    taskType,
    status,
    statusLabel: STATUS_LABELS[status] || status,
    statusClass: status === 'succeeded' ? 'green' : status === 'failed' ? 'red' : 'amber',
    provider: item.model || item.provider || 'Unknown',
    promptVersion: item.prompt_version || item.promptVersion || 'Unknown',
    outputText: `${toNumber(item.total_tokens ?? item.totalTokens)} tokens`,
    costAmount: toNumber(item.cost_amount ?? item.costAmount),
    costCurrency,
    costText: moneyText(item.cost_amount ?? item.costAmount, costCurrency),
  };
}

function normalizeRiskEvent(item = {}) {
  const riskLevel = item.risk_level || item.riskLevel || 'Unknown';
  return {
    id: item.id || '',
    taskId: item.task_id || item.taskId || null,
    channel: item.channel || 'Unknown',
    riskLevel,
    riskLabel: RISK_LABELS[riskLevel] || riskLevel,
    severity: item.severity || 'unknown',
    resolutionStatus: item.resolution_status || item.resolutionStatus || 'open',
    eventType: item.event_type || item.eventType || 'unknown',
    blockReason: item.block_reason || item.blockReason || '未填写阻断原因',
    pauseSuggested: Boolean(item.pause_suggested ?? item.pauseSuggested),
    createdAt: item.created_at || item.createdAt || '',
    highlightClass: riskClass(riskLevel),
  };
}

export function buildPhase2DashboardQuery({ channelPrefix } = {}) {
  const params = new URLSearchParams();
  if (channelPrefix) {
    params.set('channel_prefix', channelPrefix);
  }
  const query = params.toString();
  return query ? `?${query}` : '';
}

export function buildPhase2DashboardView(payload = {}) {
  const summary = payload.summary || {};
  const reviewBacklog = payload.review_backlog || {};
  const extractionStatus = payload.extraction_status_distribution || {};
  const failureReasons = Array.isArray(payload.failure_reasons) ? payload.failure_reasons : [];
  const llmCosts = payload.llm_costs || {};
  const currency = llmCosts.currency || 'CNY';
  const sourceCandidateCount = toNumber(summary.source_candidate_count ?? summary.sourceCandidateCount);
  const extractableSourceCount = toNumber(summary.auto_extraction_count ?? summary.autoExtractionCount);
  const highReviewBacklogCount = toNumber(reviewBacklog.high_risk_review ?? reviewBacklog.highRiskReview);
  const failedTaskCount = toNumber(summary.failed_task_count ?? summary.failedTaskCount);
  const agentTaskCount = toNumber(summary.agent_task_count ?? summary.agentTaskCount);
  const schemaFailureCount = failureReasons
    .filter((item) => String(item.reason || '').includes('schema'))
    .reduce((total, item) => total + toNumber(item.count), 0);
  const schemaFailureRate = agentTaskCount > 0 ? schemaFailureCount / agentTaskCount : 0;
  const llmTaskRuns = Array.isArray(llmCosts.items)
    ? llmCosts.items.map((item) => normalizeTaskRun(item, currency))
    : [];
  const highForbiddenRiskEvents = Array.isArray(payload.high_forbidden_risk_events)
    ? payload.high_forbidden_risk_events.map(normalizeRiskEvent)
    : [];

  return {
    summary: {
      sourceCandidateCount,
      extractableSourceCount,
      highReviewBacklogCount,
      reviewBacklogCount: toNumber(summary.review_backlog_count ?? summary.reviewBacklogCount),
      failedTaskCount,
      agentTaskCount,
      llmCostTotal: toNumber(summary.llm_cost_total ?? summary.llmCostTotal ?? llmCosts.total_cost),
      llmCostText: moneyText(summary.llm_cost_total ?? summary.llmCostTotal ?? llmCosts.total_cost, currency),
      riskEventCount: toNumber(summary.risk_event_count ?? summary.riskEventCount),
      highForbiddenRiskEventCount: toNumber(
        summary.high_forbidden_risk_event_count ?? summary.highForbiddenRiskEventCount,
      ),
    },
    taskFlow: [
      {
        title: 'Source Discovery',
        description: '自动发现候选来源，写入来源候选池并保留证据。',
        metricText: `${sourceCandidateCount} 来源候选`,
      },
      {
        title: 'Candidate Pool',
        description: '来源进入 lead_source_candidates，等待风险规则和人工复核。',
        metricText: `${toNumber(reviewBacklog.pending)} 待审`,
      },
      {
        title: 'Human Review',
        description: 'High 人工审核，Forbidden 阻断，不进入自动抽取。',
        metricText: `${highReviewBacklogCount} High 待审`,
      },
      {
        title: 'Extraction',
        description: 'approved 来源进入 LEAD_EXTRACTION 任务流。',
        metricText: `${extractableSourceCount} 可抽取`,
      },
      {
        title: 'Staging/Core',
        description: '抽取后进入 staging 校验、去重、复核，再进入 core。',
        metricText: `${toNumber(extractionStatus.succeeded)} 已成功`,
      },
    ],
    pauseThresholds: {
      failedTasks: {
        label: '任务失败数',
        current: failedTaskCount,
        limit: 3,
        percent: percentOf(failedTaskCount, 3),
        className: failedTaskCount >= 3 ? 'red' : 'amber',
      },
      highReviewBacklog: {
        label: 'High 审核积压',
        current: highReviewBacklogCount,
        limit: 50,
        percent: percentOf(highReviewBacklogCount, 50),
        className: highReviewBacklogCount >= 50 ? 'red' : 'amber',
      },
      schemaFailureRate: {
        label: 'schema 失败率',
        current: schemaFailureRate,
        limit: 0.05,
        percent: Math.min(100, Math.round(schemaFailureRate * 100)),
        text: `${(schemaFailureRate * 100).toFixed(1)}%`,
        className: schemaFailureRate >= 0.05 ? 'red' : 'green',
      },
    },
    llmTaskRuns,
    failureReasons,
    highForbiddenRiskEvents,
    guardrail: payload.guardrail || '第二阶段运行看板必须遵守风险边界和审计要求。',
  };
}

export async function fetchPhase2Dashboard({
  baseUrl = '',
  channelPrefix,
  fetcher = globalThis.fetch,
} = {}) {
  if (typeof fetcher !== 'function') {
    throw new Error('fetcher is required to load phase2 dashboard');
  }

  const normalizedBaseUrl = String(baseUrl || '').replace(/\/$/, '');
  const response = await fetcher(`${normalizedBaseUrl}/dashboard/phase2${buildPhase2DashboardQuery({ channelPrefix })}`);
  if (!response.ok) {
    throw new Error(`Failed to load phase2 dashboard: ${response.status || 'unknown'}`);
  }
  return response.json();
}

File: apps/api/app/services/phase2_dashboard.py
from __future__ import annotations

from collections import Counter, defaultdict
from typing import Any

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models.agent_task_run import AgentTaskRun
from app.models.enums import (
    AgentTaskRunStatus,
    AgentTaskType,
    ChannelRiskLevel,
    LeadSourceCandidateExtractionStatus,
    LeadSourceCandidateReviewStatus,
    RiskEventStatus,
)
from app.models.lead_source_candidate import LeadSourceCandidate
from app.models.risk_event import RiskEvent


class Phase2DashboardService:
    GUARDRAIL = (
        "第二阶段 dashboard 仅用于运行监控；不自动社交私信、不自动加好友、不登录后批量采集、"
        "不反爬规避；High/Forbidden 风险必须单独复核。"
    )

    def __init__(self, session: Session) -> None:
        self.session = session

    def metrics(self, *, channel_prefix: str | None = None) -> dict[str, Any]:
        candidates = self._load_candidates(channel_prefix=channel_prefix)
        task_runs = self._load_task_runs(channel_prefix=channel_prefix)
        risk_events = self._load_risk_events(channel_prefix=channel_prefix)

        risk_distribution = self._distribution(
            [candidate.risk_level.value for candidate in candidates],
            [risk.value for risk in ChannelRiskLevel],
        )
        review_backlog = self._review_backlog(candidates)
        extraction_distribution = self._distribution(
            [candidate.extraction_status.value for candidate in candidates],
            [status.value for status in LeadSourceCandidateExtractionStatus],
        )
        failure_reasons = self._failure_reasons(task_runs)
        llm_costs = self._llm_costs(task_runs)
        high_forbidden_risk_events = [
            self._serialize_risk_event(event)
            for event in risk_events
            if event.risk_level in {ChannelRiskLevel.HIGH, ChannelRiskLevel.FORBIDDEN}
        ]

        return {
            "summary": {
                "source_candidate_count": len(candidates),
                "review_backlog_count": sum(review_backlog.values()),
                "auto_extraction_count": sum(
                    1 for candidate in candidates if candidate.extraction_status == LeadSourceCandidateExtractionStatus.SUCCEEDED
                ),
                "agent_task_count": len(task_runs),
                "failed_task_count": sum(1 for task in task_runs if task.status == AgentTaskRunStatus.FAILED),
                "llm_cost_total": llm_costs["total_cost"],
                "risk_event_count": len(risk_events),
                "high_forbidden_risk_event_count": len(high_forbidden_risk_events),
            },
            "risk_distribution": risk_distribution,
            "review_backlog": dict(review_backlog),
            "extraction_status_distribution": extraction_distribution,
            "failure_reasons": failure_reasons,
            "llm_costs": llm_costs,
            "high_forbidden_risk_events": high_forbidden_risk_events,
            "guardrail": self.GUARDRAIL,
        }

    def _load_candidates(self, *, channel_prefix: str | None) -> list[LeadSourceCandidate]:
        statement = select(LeadSourceCandidate)
        if channel_prefix:
            statement = statement.where(LeadSourceCandidate.channel_name.like(f"{channel_prefix}%"))
        return list(self.session.scalars(statement).all())

    def _load_task_runs(self, *, channel_prefix: str | None) -> list[AgentTaskRun]:
        statement = select(AgentTaskRun).where(
            AgentTaskRun.task_type.in_(
                [
                    AgentTaskType.SOURCE_DISCOVERY,
                    AgentTaskType.LEAD_EXTRACTION,
                    AgentTaskType.LEAD_GRADING,
                    AgentTaskType.RETRY_WORKER,
                ]
            )
        )
        if channel_prefix:
            statement = statement.where(AgentTaskRun.trigger_source.like(f"{channel_prefix}%"))
        return list(self.session.scalars(statement).all())

    def _load_risk_events(self, *, channel_prefix: str | None) -> list[RiskEvent]:
        statement = select(RiskEvent).order_by(RiskEvent.created_at.desc())
        if channel_prefix:
            statement = statement.where(RiskEvent.channel.like(f"{channel_prefix}%"))
        return list(self.session.scalars(statement).all())

    def _review_backlog(self, candidates: list[LeadSourceCandidate]) -> dict[str, int]:
        backlog_statuses = {
            LeadSourceCandidateReviewStatus.PENDING,
            LeadSourceCandidateReviewStatus.HIGH_RISK_REVIEW,
            LeadSourceCandidateReviewStatus.NEEDS_RECHECK,
        }
        counter: Counter[str] = Counter()
        for candidate in candidates:
            if candidate.review_status in backlog_statuses:
                counter[candidate.review_status.value] += 1
        return dict(counter)

    def _failure_reasons(self, task_runs: list[AgentTaskRun]) -> list[dict[str, Any]]:
        grouped: dict[str, list[str]] = defaultdict(list)
        for task in task_runs:
            if task.status != AgentTaskRunStatus.FAILED:
                continue
            reason = self._failure_reason(task)
            grouped[reason].append(str(task.id))
        return [
            {"reason": reason, "count": len(ids), "agent_task_run_ids": ids}
            for reason, ids in sorted(grouped.items(), key=lambda item: (-len(item[1]), item[0]))
        ]

    def _failure_reason(self, task: AgentTaskRun) -> str:
        summary = task.output_summary_json if isinstance(task.output_summary_json, dict) else {}
        error = summary.get("error") if isinstance(summary.get("error"), dict) else {}
        return str(error.get("type") or task.error_message or "unknown_failure")

    def _llm_costs(self, task_runs: list[AgentTaskRun]) -> dict[str, Any]:
        items = []
        total_cost = 0.0
        currency = "USD"
        for task in task_runs:
            summary = task.output_summary_json if isinstance(task.output_summary_json, dict) else {}
            cost_amount = float(summary.get("cost_amount") or 0)
            token_usage = task.token_usage_json if isinstance(task.token_usage_json, dict) else {}
            total_tokens = int(token_usage.get("total_tokens") or 0)
            if cost_amount <= 0 and total_tokens <= 0:
                continue
            cost_currency = str(summary.get("cost_currency") or currency)
            currency = cost_currency
            total_cost += cost_amount
            items.append(
                {
                    "agent_task_run_id": str(task.id),
                    "task_type": task.task_type.value,
                    "status": task.status.value,
                    "model": task.llm_model,
                    "prompt_version": task.prompt_version,
                    "cost_amount": cost_amount,
                    "cost_currency": cost_currency,
                    "total_tokens": total_tokens,
                }
            )
        return {"total_cost": round(total_cost, 4), "currency": currency, "items": items}

    def _serialize_risk_event(self, event: RiskEvent) -> dict[str, Any]:
        return {
            "id": str(event.id),
            "task_id": event.task_id,
            "channel": event.channel,
            "risk_level": event.risk_level.value,
            "severity": event.severity.value,
            "resolution_status": event.resolution_status.value if isinstance(event.resolution_status, RiskEventStatus) else str(event.resolution_status),
            "event_type": event.event_type,
            "block_reason": event.block_reason,
            "pause_suggested": event.pause_suggested,
            "created_at": event.created_at.isoformat(),
        }

    def _distribution(self, values: list[str], keys: list[str]) -> dict[str, int]:
        counter = Counter(values)
        return {key: counter.get(key, 0) for key in keys if counter.get(key, 0) > 0}
